resource_work_time_from_contracts/models/resource_mixin.py

```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
+        return self._sum_intervals(work_intervals)
+
+    def list_normal_work_time_per_day(self, from_datetime, to_datetime):
+        """
+        Same as list_work_time_per_day(), but ignoring leaves
+        """
+        return self._get_work_time_from_contracts(from_datetime, to_datetime)
+
+    def get_attendances_of_date_range(self, date_from, date_to):
+        """
+        Get the first and last resource.calendar.attendance corresponding to
+        the range defined by date_from and date_to.
+        """
+        contracts = self._get_active_contracts(date_from, date_to)
+        if not contracts:
+            (
+                first_attendance,
+                last_attendance,
+            ) = self.resource_calendar_id.get_first_last_attendance(date_from, date_to)
+            if first_attendance is None:
+                return (None, None)
+            return (first_attendance[1], last_attendance[1])
+        earliest_attendance = None
+        latest_attendance = None
+        for contract in contracts:
+            (
+                attendance_from,
+                attendance_to,
+            ) = contract.resource_calendar_id.get_first_last_attendance(
+                max(date_from, contract.date_start),
+                min(date_to, contract.date_end or date_to),
+            )
+            if attendance_from is not None and (
+                earliest_attendance is None
+                or attendance_from[0] < earliest_attendance[0]
+                or (
+                    attendance_from[0] == earliest_attendance[0]
+                    and attendance_from[1].hour_from < earliest_attendance[1].hour_from
+                )
+            ):
+                earliest_attendance = attendance_from
+            if attendance_to is not None and (
+                latest_attendance is None
+                or attendance_to[0] > latest_attendance[0]
+                or (
+                    attendance_to[0] == latest_attendance[0]
+                    and attendance_to[1].hour_to > latest_attendance[1].hour_to
+                )
+            ):
+                latest_attendance = attendance_to
+        if earliest_attendance is None:
+            return (None, None)
+        return (earliest_attendance[1], latest_attendance[1])
+
+    def get_calendar_for_date(self, date):
+        """
+        Get the resource.calendar of the contract that applies to the provided
+        date.
+        """
+        # there can be multiple contracts active at the same time. only the
+        # first one found will be used.
+        current_contracts = self._get_active_contracts(date, date)
+        if not current_contracts:
+            return None
+        return current_contracts[0].resource_calendar_id
+
+    def _sum_intervals(self, intervals):
+        result = defaultdict(float)
+        for start, stop, meta in intervals:
+            result[start.date()] += (stop - start).total_seconds() / 3600
+        return sorted(result.items())
```

Comment:
Bad method name. `_hours_per_date[_from_intervals]` would work better.

Reply:
In fact, `_hours_per_day_from_intervals` already exists… It appears to do something frustratingly adjacent, but subtly different.

Reply:
renamed `sum_intervals()` to `_sum_hours_per_date_from_intervals()` and `_hours_per_day_from_intervals()` to `_expected_attendance_per_day_from_intervals()`.